src/backend/distributed/commands/type.c

```diff
@@ -0,0 +1,865 @@
+		 * might already be added to some nodes, but not all.
+		 */
+
+		/* TODO function name is unwieldly long, and runs serially which is not nice */
+		List *commands = list_make2(DISABLE_DDL_PROPAGATION, (void *) alterEnumStmtSql);
+		int result =
+			SendBareOptionalCommandListToWorkersAsUser(ALL_WORKERS, commands, NULL);
+
+		if (result != RESPONSE_OKAY)
+		{
+			const char *alterEnumStmtIfNotExistsSql = NULL;
+			bool oldSkipIfNewValueExists = stmt->skipIfNewValExists;
+
+			/* deparse the query with IF NOT EXISTS */
+			stmt->skipIfNewValExists = true;
+			alterEnumStmtIfNotExistsSql = deparse_alter_enum_stmt(stmt);
+			stmt->skipIfNewValExists = oldSkipIfNewValueExists;
+
+			ereport(WARNING, (errmsg("not all workers applied change to enum"),
+							  errdetail("retry with: %s", alterEnumStmtIfNotExistsSql),
+							  errhint("make sure the coordinators can communicate with "
+									  "all workers")));
+		}
+	}
+	else
+	{
+		/* other statements can be run in a transaction and will be dispatched here. */
+		EnsureSequentialModeForTypeDDL();
+		SendCommandToWorkersAsUser(ALL_WORKERS, DISABLE_DDL_PROPAGATION, NULL);
+		SendCommandToWorkersAsUser(ALL_WORKERS, alterEnumStmtSql, NULL);
+	}
+
+	return NIL;
+}
+
+
+/*
+ * PlanDropTypeStmt is called for all DROP TYPE statements. For all types in the list that
+ * citus has distributed to the workers it will drop the type on the workers as well. If
+ * no types in the drop list are distributed no calls will be made to the workers.
+ */
+List *
+PlanDropTypeStmt(DropStmt *stmt, const char *queryString)
+{
+	/*
+	 * We swap the list of objects to remove during deparse so we need a reference back to
+	 * the old list to put back
+	 */
+	List *oldTypes = stmt->objects;
+	List *distributedTypes = NIL;
+	const char *dropStmtSql = NULL;
+	ListCell *addressCell = NULL;
+	List *distributedTypeAddresses = NIL;
+
+	if (creating_extension)
+	{
+		/*
+		 * extensions should be created separately on the workers, types cascading from an
+		 * extension should therefor not be propagated here.
+		 */
+		return NIL;
+	}
+
+	/*
+	 * managing types can only be done on the coordinator if ddl propagation is on. when
+	 * it is off we will never get here. MX workers don't have a notion of distributed
+	 * types, so we block the call.
+	 */
+	EnsureCoordinator();
+
+	distributedTypes = FilterNameListForDistributedTypes(oldTypes);
+	if (list_length(distributedTypes) <= 0)
+	{
+		/* no distributed types to drop */
+		return NULL;
+	}
+
+	/*
+	 * temporary swap the lists of objects to delete with the distributed objects and
+	 * deparse to an executable sql statement for the workers
+	 */
+	stmt->objects = distributedTypes;
+	dropStmtSql = deparse_drop_type_stmt(stmt);
+	stmt->objects = oldTypes;
+
+	/* to prevent recursion with mx we disable ddl propagation */
+	EnsureSequentialModeForTypeDDL();
+	SendCommandToWorkersAsUser(ALL_WORKERS, DISABLE_DDL_PROPAGATION, NULL);
```

Comment:
One side effect of running the commands on the shards before the coordinator is that the errors comes from the worker node. For example, all DDLs first executed on the coordinator. If anything is wrong, the error is thrown. Else, we'll run the commands on the shards as well.

```SQL
CREATE TYPE drop_type_test AS (user_name text, user_id bigint);
CREATE TABLE drop_type_table (a int, b drop_type_test);
SELECT create_distributed_table('drop_type_table', 'a');

 DROP TYPE  drop_type_test;

ERROR:  cannot drop type drop_type_test because other objects depend on it
DETAIL:  column b of table drop_type_table_102623 depends on type drop_type_test
column b of table drop_type_table_102625 depends on type drop_type_test
column b of table drop_type_table_102627 depends on type drop_type_test
column b of table drop_type_table_102629 depends on type drop_type_test
column b of table drop_type_table_102631 depends on type drop_type_test
column b of table drop_type_table_102633 depends on type drop_type_test
column b of table drop_type_table_102635 depends on type drop_type_test
column b of table drop_type_table_102637 depends on type drop_type_test
column b of table drop_type_table_102639 depends on type drop_type_test
column b of table drop_type_table_102641 depends on type drop_type_test
column b of table drop_type_table_102643 depends on type drop_type_test
column b of table drop_type_table_102645 depends on type drop_type_test
column b of table drop_type_table_102647 depends on type drop_type_test
column b of table drop_type_table_102649 depends on type drop_type_test
column b of table drop_type_table_102651 depends on type drop_type_test
column b of table drop_type_table_102653 depends on type drop_type_test
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
CONTEXT:  while executing command on localhost:9700
Time: 21.078 ms
```

We've this behaviour in other places as well such as `alter type test_enum add value ` (and probably others as well)


Also, the function name starts with `Plan..`, and we also execute it.

Reply:
Yeah, would be great to actually 'plan' for the drop statements.
All other calls seem to rely on the changes be visible locally before constructing the sql to propagate as they sometimes rely on oid's to resolve correctly.

For drop we actually need to construct the sql before we execute locally as the type is not visible anymore after the drop, which causes names to not resolve correctly.

Problem I had with planning ddl jobs is that they seemed to be very shard centric, will have a new look at it to see if we can use it.

I agree that names should reflect correctly in any case.

Reply:
This is solved now that the executor is used